src/framework/mpas_constants.F

```diff
@@ -22,35 +22,57 @@ module mpas_constants
 
    use mpas_kind_types
 
+#ifdef MPAS_CAM_DYCORE
+   use physconst, only : pii => pi
+   use physconst, only : gravity => gravit
+   use physconst, only : omega
+   use physconst, only : a => rearth
+   use physconst, only : cp => cpair
+   use physconst, only : rgas => rair
+   use physconst, only : rv => rh2o
+   real (kind=RKIND) :: rvord = huge(1.0_RKIND)                       ! Derived in mpas_constants_compute_derived
+   real (kind=RKIND) :: cv = huge(1.0_RKIND)                          ! Derived in mpas_constants_compute_derived
+   real (kind=RKIND) :: cvpm = huge(1.0_RKIND)                        ! Derived in mpas_constants_compute_derived
+#else
    real (kind=RKIND), parameter :: pii     = 3.141592653589793_RKIND  !< Constant: Pi
    real (kind=RKIND), parameter :: a       = 6371229.0_RKIND          !< Constant: Spherical Earth radius [m]
    real (kind=RKIND), parameter :: omega   = 7.29212e-5_RKIND         !< Constant: Angular rotation rate of the Earth [s-1]
    real (kind=RKIND), parameter :: gravity = 9.80616_RKIND            !< Constant: Acceleration due to gravity [m s-2]
    real (kind=RKIND), parameter :: rgas    = 287.0_RKIND              !< Constant: Gas constant for dry air [J kg-1 K-1]
    real (kind=RKIND), parameter :: rv      = 461.6_RKIND              !< Constant: Gas constant for water vapor [J kg-1 K-1]
-   real (kind=RKIND), parameter :: rvord   = rv / rgas                !
 !  real (kind=RKIND), parameter :: cp      = 1003.0_RKIND             !< Constant: Specific heat of dry air at constant pressure [J kg-1 K-1]
    real (kind=RKIND), parameter :: cp      = 7.0_RKIND*rgas/2.0_RKIND !< Constant: Specific heat of dry air at constant pressure [J kg-1 K-1]
+   real (kind=RKIND), parameter :: rvord   = rv / rgas                !
    real (kind=RKIND), parameter :: cv      = cp - rgas                !< Constant: Specific heat of dry air at constant volume [J kg-1 K-1]
    real (kind=RKIND), parameter :: cvpm    = -cv / cp                 !
+#endif
+   real (kind=RKIND), parameter :: p0 = 1.0e5_RKIND                   !< Constant: 100000 Pa
    real (kind=RKIND), parameter :: prandtl = 1.0_RKIND                !< Constant: Prandtl number
 
    contains
 
 
 !***********************************************************************
 !
-!  routine dummy
+!  mpas_constants_compute_derived
 !
-!> \brief   MPAS Dummy Routine
+!> \brief   Computes derived constants
 !> \author  Michael Duda
-!> \date    03/27/13
+!> \date    8 May 2020
 !> \details 
-!> This is a dummy routine that doesn't do anything.
+!>  This routine computes constants that are derived from others.
 !
 !-----------------------------------------------------------------------
-   subroutine dummy()
+   subroutine mpas_constants_compute_derived()
+
+       implicit none
+
+#ifdef MPAS_CAM_DYCORE
+       rvord   = rv / rgas
+       cv      = cp - rgas
+       cvpm    = -cv / cp
```

Comment:
Is it not possible to use variables from a use statement in the deceleration section? For example, could we duplicate line 45 to line 33 where CAM-MPAS constants are defined?

Reply:
The reason why we do this isn't obvious. Essentially, some of these constants (in a physical sense, not so much a code sense) can change based on namelist options, so we can only compute some physical constants later on in the initialization process. Take `rair` (which we import as `rgas`) for example: it may be modified at runtime by the `physconst_readnl` routine, so we can only compute physical constants that depend on `rair` after that modification point. Here's what the relevant parts of the call tree look like in CAM-MPAS:
```
cam_init
   read_namelist
      physconst_readnl
   dyn_init
      mpas_constants_compute_derived
```